server/src/user/middleware.ts

```diff
@@ -27,63 +31,172 @@ export class Middleware {
+            const cookies = cookie.parse(socket.request.headers.cookie || '');
+            if(!cookies.authenticationToken) {
+                // websocket connection requires an authentication token
+                // TODO: restructure request logs to make message more human readable, rather than a separate 'reason' message
+                GlobalLogger.request('ws-handshake-failed', {
+                    ip: socket.request.connection.remoteAddress,
+                    transport: 'ws',
+                    reason: 'missing authenticationToken'
+                })
+                return next(new UnauthorizedError(`Websocket connection requires a the 'authenticationToken' cookie. Obtain a token with POST /users/login`));
+            }
+
+            this.service.verifyAuthenticationToken(cookies.authenticationToken).pipe(
+                catchError((err) => throwError(new UnauthorizedError(err.message)))
+            ).subscribe({
+                next: (authenticationToken) => {
+                    socket.mazenet = {
+                        sessionId: socket.id,
+                        userId: authenticationToken.sub
+                    };
 
-            socket.on('disconnect', () => {
-                this.service.onUserDisconnect(socket.mazenet!.sessionId);
+                    socket.on('disconnect', () => {
+                        this.service.onUserDisconnect(socket.mazenet!.sessionId);
+                    });
+                    next();
+                },
+                error: (error: Error) => {
+                    GlobalLogger.error(
+                        'Socket failed authentication handshake on connect',
+                        {error}
+                    );
+                    next(error);
+                }
             });
-            next();
         };
         this.router = Express.Router();
+
         this.router.use((req: Request, res: Response, next: Express.NextFunction) => {
             const socketData = (req.socket as Socket).mazenet;
             if(socketData) {
-                // this is a socketio connection
-                //TODO: do we need to validate that activeUser is valid for this user?
-                //TODO: standardize error handling
-                req.activeUser = this.service.getActiveUserFromSession(socketData!.sessionId);
-                if(!socketData!.user) {
-                    this.service.createUser({username: 'anonymous'}).subscribe((user) => {
-                        socketData!.user = user;
-                        req.user = user;
-                        return next();
-                    }, (error) => next(error));
-                } else {
-                    req.user = socketData!.user;
-                    return next();
+                // websocket doesn't need to validate JWT because it was validated when the connection was established
+                // NOTE: should we kill long-running websocket connections that have outlived their token lifespan?
```

Comment:
Killing the connection on client-side wouldn't be secure. We should store the expiration time in the socket data on connection, and check on each request if the current time has exceeded the expiration.

Reply:
Oh I misunderstood. I thought the socket would just start erroring when the cert expired, but I get it now that it would continue as an authenticated pipe forever.

Reply:
What really should happen is we provide a token refresh mechanism, where the client detects that its session will expire soon and automatically requests a new session token. If the client fails to do this in time, the sever closes the connection when the session expires.

For reference, here's how OpenID does refresh tokens:
 - On authentication, you can request to get a refresh_token in addition to the id_token.
 - The user securely stores the refresh_token (e.g. encrypted in phone local storage).
 - The refresh_token has a long lifespan, and is only used infrequently to obtain a new session token. The session token cannot be used to refresh itself.
 - The refresh_token can only be renewed through a full authentication flow.

The big problem with automatically refreshing tokens is that if your refresh token is stolen by an adversary, they can refresh the stolen token and retain access to your account for the lifetime of the refresh token (usually 1-3 months). These days, many websites allow you to refresh your refresh_token, which opens the possibility of an adversary having access to your account indefinitely. It seems like something we've accepted as a trade-off on internet authentication in general for the convenience.